### PR TITLE
Log updates

### PR DESCRIPTION
**Pull request by moldover**
Some much needed refactoring of Logging to fix "double console logging" but more importantly, greatly simplify logging to use multiple LogHandlers instead of multiple Loggers.  I tested this with a few tabulations from the CLI and GUI and it seems ok, but if you guys wouldn't mind running a couple tabulations and verifying the logging outputs look ok that would be greatly appreciated.  

Users of the Logger no longer need to know about different flavors of logger.  They DO need to be aware of the LEVEL they use, notably: Level.FINE will go into the execution log only (for audit logging).  Level.INFO (and above) will go into execution log file, tabulation file (if tabulating) and GUI (if GUI mode).  So for most use cases, you just call Logger.Log(Level.X, ...) and get on with your day :)

Note: this does not attempt to address the issue of potentially huge audit log files.  @CalebKleppner: it would be fairly easy to split up a huge audit output (using the Logger framework) say every 100MB (or 50, etc.) gets written to a different file.  Is that something that would be useful?

### CODE REVIEW

File: src/main/java/com/rcv/Logger.java [hunk 1]
```diff
@@ -44,145 +53,110 @@
...
-    consoleHandler.setFormatter(formatter);
-    consoleHandler.setLevel(Level.FINE);
-    // add the handlers
-    executionLogger.addHandler(consoleHandler);
+        new FileHandler(logPath.toString(),
+            EXECUTION_LOG_MAX_SIZE_BYTES,
+            EXECUTION_LOG_FILE_COUNT,
+            true);
+    // use our custom formatter
+    fileHandler.setFormatter(new LogFormatter());
```
Comment (HEdingfield) [inline src/main/java/com/rcv/Logger.java]: Might make more sense and be more memory-efficient to just instantiate a `static final` of the formatter and use it for all.

File: src/main/java/com/rcv/Logger.java [hunk 2]
```diff
@@ -44,145 +53,110 @@
...
   // execution log file name (%g tracks count of log file if additional versions are created)
   private static final String EXECUTION_LOG_FILE_NAME = "rcv_%g.log";
   // first value here is bytes per MB and the second is max MB for the log file
-  private static final Integer EXECUTION_LOG_FILE_MAX_SIZE_BYTES = 1000000 * 50;
+  private static final Integer EXECUTION_LOG_MAX_SIZE_BYTES = 1000000 * 50;
   // how many execution files to keep
   private static final Integer EXECUTION_LOG_FILE_COUNT = 2;
   // cache for the execution logger
   private static java.util.logging.Logger executionLogger;
   // cache for the tabulation logger
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Logger.java]: sure

File: src/main/java/com/rcv/Logger.java [hunk 3]
```diff
@@ -44,145 +53,110 @@
...
-  private static final String GUI_LOGGER_NAME = "GUI";
   // execution log file name (%g tracks count of log file if additional versions are created)
   private static final String EXECUTION_LOG_FILE_NAME = "rcv_%g.log";
   // first value here is bytes per MB and the second is max MB for the log file
-  private static final Integer EXECUTION_LOG_FILE_MAX_SIZE_BYTES = 1000000 * 50;
+  private static final Integer EXECUTION_LOG_MAX_SIZE_BYTES = 1000000 * 50;
   // how many execution files to keep
   private static final Integer EXECUTION_LOG_FILE_COUNT = 2;
   // cache for the execution logger
   private static java.util.logging.Logger executionLogger;
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Logger.java]: I'm updating the comments and variables.  this is (still) the execution log file (not logger) etc. so I think the name is correct.

File: src/main/java/com/rcv/Logger.java [hunk 4]
```diff
@@ -44,145 +53,110 @@
...
     // "user.dir" property is the current working directory, i.e. folder from whence the rcv jar
     // was launched
-    String logPath = Paths.get(System.getProperty("user.dir"), EXECUTION_LOG_FILE_NAME).toString();
-    // formatter specifies how logging output lines should appear
-    LogFormatter formatter = new LogFormatter();
+    Path logPath = Paths.get(System.getProperty("user.dir"),
+        EXECUTION_LOG_FILE_NAME).toAbsolutePath();
+
     // fileHandler writes formatted strings to file
     FileHandler fileHandler =
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Logger.java]: sure.  drastic?

File: src/main/java/com/rcv/Main.java [hunk 5]
```diff
@@ -96,36 +96,35 @@ static ContestConfig loadContestConfig(String configPath) {
...
   static void executeTabulation(ContestConfig config) {
-    Logger.executionLog(Level.INFO, "Starting tabulation process...");
+    Logger.Log(Level.INFO, "Starting tabulation process...");
 
     boolean isTabulationCompleted = false;
     boolean isConfigValid = config.validate();
 
     if (isConfigValid) {
-      boolean isTabulationLogSetUp = false;
+      boolean isexecutionLogSetUp = false;
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Main.java]: whoops!

File: src/main/java/com/rcv/Logger.java (hunk 1 above)
Comment (moldover) [reply src/main/java/com/rcv/Logger.java]: dig

File: src/main/java/com/rcv/Logger.java [hunk 6]
```diff
@@ -44,145 +53,110 @@
...
+      @Override
+      public void flush() {
+      }
+      // nothing to do here
+      @Override
+      public void close() {
+      }
+    };
+    guiHandler.setLevel(Level.INFO);
+    guiHandler.setFormatter(new LogFormatter());
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Logger.java]: dig

File: src/main/java/com/rcv/Main.java [hunk 7]
```diff
@@ -96,36 +96,35 @@ static ContestConfig loadContestConfig(String configPath) {
...
       // current date-time formatted as a string used for creating unique output files names
       final String timestampString = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss").format(new Date());
-      String tabulationLogPath =
-          Paths.get(config.getOutputDirectory(), String.format("%s_audit.log", timestampString))
-              .toString();
-
+      String executionLogPath =
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Main.java]: yep

File: src/main/java/com/rcv/Logger.java [hunk 8]
```diff
@@ -42,147 +52,111 @@
...
-          @Override
-          public void flush() {
-          }
-
-          @Override
-          public void close() {
-          }
-        });
+    // log results
+    log(Level.INFO,"RCV Tabulator logging execution to %s", logPath.toString());
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/Logger.java]: %g isn't a String format code; FileHandler uses it, so this is "correct".  It's just debug output so I think it's fine.  If we ever show it to the user I'll clean it up somehow.

**Comment by CalebKleppner** [comment]
Yes, it would make sense to split big output files into chunks of, say, 
50MB.  I think there's a bunch of room for improvements here, both in 
terms of size of auditing files and in terms of readability and 
usability of the files.  I'm not sure what the best answer is, but we 
may want to think about 3 separate audit logs:

1. The narrative of each round:  how many votes each candidate had, who 
got elected, eliminated, etc.
2. How each ballot was counted in first round.  Would include full raw 
data of each CVR (including all ballot IDs, precinct IDs and any other 
ballot headers that might be included)
3. Record of transfers (for round 2 to n):  ballot ID, transferred from, 
transferred to, transfer weight

I'm attaching the input and output data from 2017 Cambridge city council 
election. STV election for 9 seats using whole transfers. The input 
files are 33 .prm files containing rankings using codes and a .chp file, 
which is the election configuration (candidates, rules, CVR sources).  
The outputs are the .html files (round by round results) and a bunch of 
auditing reports (ballot allocation report showing how each ballot 
counted in the final round, tabulator report showing a table of results 
by round, transfer report, and a few more).

Ultimately, we want to be able to tell how each ballot was counted in 
each round, which our current log file does very well, and we want to 
see what happens in a round (ie, all the transfers in that round), which 
our current log doesn't do so easily.

Since this doesn't affect the tabulation, we don't need to improve the 
log files until we're done with other things.

Best,
Caleb

Caleb Kleppner
MK Election Services, LLC
203-376-4080

**Pull request merged**